src/Adapter/Swoole/EventLoop.php

```diff
@@ -0,0 +1,282 @@
+<?php
+
+namespace M6Web\Tornado\Adapter\Swoole;
+
+use M6Web\Tornado\Adapter\Common\Internal\FailingPromiseCollection;
+use M6Web\Tornado\Adapter\Swoole\Internal\SwooleDeferred;
+use M6Web\Tornado\Adapter\Swoole\Internal\SwoolePromise;
+use M6Web\Tornado\Deferred;
+use M6Web\Tornado\Promise;
+use Swoole\Coroutine;
+use Swoole\Coroutine\Channel;
+use Swoole\Coroutine\WaitGroup;
+use Swoole\Event;
+
+class EventLoop implements \M6Web\Tornado\EventLoop
+{
+    /** @var Internal\StreamEventLoop */
+    private $streamLoop;
+
+    /** @var FailingPromiseCollection */
+    private $unhandledFailingPromises;
+
+    public function __construct()
+    {
+        $this->streamLoop = new Internal\StreamEventLoop();
+        $this->unhandledFailingPromises = new FailingPromiseCollection();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function wait(Promise $promise)
+    {
+        $value = null;
+        $isRejected = false;
+        $promiseSettled = false;
+
+        Coroutine::create(function() use ($promise, &$value, &$isRejected, &$promiseSettled) {
+            $ticks = 1;
+            //$wg = new WaitGroup(1);
+            $channel = new Channel($ticks);
+            Internal\PromiseWrapper::toHandledPromise($promise, $this->unhandledFailingPromises)->getSwoolePromise()->then(
+                function ($result) use (/* $wg , */$channel, &$value, &$promiseSettled) {
+                    $promiseSettled = true;
+                    $value = $result;
+                    $channel->push(true);
+                    //$wg->done();
+                },
+                function ($result) use (/* $wg , */$channel, &$value, &$isRejected, &$promiseSettled) {
+                    $promiseSettled = true;
+                    $value = $result;
+                    $isRejected = true;
+                    $channel->push(true);
+                    //$wg->done();
+                }
+            );
+            while ($ticks--) {
+                $channel->pop();
+            }
+            $channel->close();
+            //$wg->wait();
+        });
+        while (!$promiseSettled) {
+            // @codeCoverageIgnoreStart
+            usleep(SwoolePromise::PROMISE_WAIT);
+            // @codeCoverageIgnoreEnd
+        }
+        //Event::wait();
+        //swoole_event_wait();
+
+        if (!$promiseSettled) {
+            throw new \Error('Impossible to resolve the promise, no more task to execute.');
+        }
+
+        if ($isRejected) {
+            /* @var $value \Exception */
+            throw $value;
+        }
+
+        $this->unhandledFailingPromises->throwIfWatchedFailingPromiseExists();
+
+        return $value;
+    }
```

Comment:
I don't see why we would need a new Coroutine here 🤔 
It seems that you copied some code from the [ReactPhp adapter](https://github.com/M6Web/Tornado/blob/e9bf5e772995f97df7868efade706eae871bf54b/src/Adapter/ReactPhp/EventLoop.php#L31-L60), looks like a good idea to me 👍 I wonder if we could just apply the same logic ⏬ 

```php
// By design, Tornado expects the resolution of a particular Promise to end the event loop.
// Since it's not the case in Swoole (nor in ReactPhp), we explicitly chain the promise resolution with an exit of the loop.
// … and we store the result :)
Internal\PromiseWrapper::toHandledPromise($promise, $this->unhandledFailingPromises)->getSwoolePromise()->then(
    function ($result) use (&$value, &$promiseSettled) {
        $promiseSettled = true;
        $value = $result;
        Swoole\Event->exit();
     },
    function ($result) use (&$value, &$isRejected, &$promiseSettled) {
        $promiseSettled = true;
        $value = $result;
        $isRejected = true;
        Swoole\Event->exit();
       }
);

// In some particular cases, the promise is already resolved, then we don't need to run the loop.
// If a promise is provided, it means that the loop has "something to do".
// Then we run it, until our promise is resolved and exits the loop.
if (!$promiseSettled) {
    Swoole\Event->wait();
}

// If our Promise is not resolved, it means that the loop has been stop before… it's unexpected!
if (!$promiseSettled) {
    throw new \Error('Impossible to resolve the promise, no more task to execute.');
}

// Rejected promise must throw something
if ($isRejected) {
    throw $value;
}

// This is Tornado specific, sometimes we launch "background" coroutines for some generic jobs,
// for example to process the GuzzleHTtp sub-event loop.
// In those cases, nobody care about corresponding promise and nobody is notified if an error occurs 😕 
// When the loop ends, we check that one of these promises raised an error.
$this->unhandledFailingPromises->throwIfWatchedFailingPromiseExists();

return $value;
```

Reply:
Swoole force to use `Coroutine::create` when using `Channel` or `WaitGroup`, if not Swoole produce an Exception : `Swoole\Error: API must be called in the coroutine`

Swoole\Event->wait seem to wait for all Swoole Coroutine in action, and this would break `testWaitFunctionShouldReturnAsSoonAsPromiseIsResolved`.

For more idea about Swoole usage, there are lots of implementation example on https://github.com/swoole/swoole-src/tree/master/examples but all seems very simple use case.

Reply:
> Swoole force to use Coroutine::create when using Channel or WaitGroup, if not Swoole produce an Exception : Swoole\Error: API must be called in the coroutine
I'm still not convinced that we need a Channel or a WaitGroup with proposed code… will have to read some documentation

> Swoole\Event->wait seem to wait for all Swoole Coroutine in action, and this would break testWaitFunctionShouldReturnAsSoonAsPromiseIsResolved.
That's exactly why I call `Swoole\Event->exit()` as soon as the input promise is resolved. I didn't test yet, but should work in my understanding… and then we don't need any Channel or Waitgroup :)

Reply:
I should try with Swoole\Event->exit();

Reply:
it seem to work !

`bin/phpunit tests/Adapter/Swoole/EventLoopTest.php --filter testWaitFunctionShouldReturnAsSoonAsPromiseIsResolved`